src/lib.rs

```diff
@@ -948,16 +952,56 @@ impl<S: VhostUserBackend> VhostUserSlaveReqHandlerMut for VhostUserHandler<S> {
 
     fn get_inflight_fd(
         &mut self,
-        _inflight: &vhost::vhost_user::message::VhostUserInflight,
-    ) -> VhostUserResult<(vhost::vhost_user::message::VhostUserInflight, File)> {
-        // Assume the backend hasn't negotiated the inflight feature; it
-        // wouldn't be correct for the backend to do so, as we don't (yet)
-        // provide a way for it to handle such requests.
-        Err(VhostUserError::InvalidOperation)
+        inflight: &VhostUserInflight,
+    ) -> VhostUserResult<(VhostUserInflight, File)> {
+        let inflight_queue_region = unsafe {
+            libc::syscall(
+                libc::SYS_memfd_create,
+                &std::ffi::CString::new("inflight-region").unwrap(),
+                libc::MFD_CLOEXEC,
+            )
+        };
+
+        self.inflight_file = Some(unsafe { File::from_raw_fd(inflight_queue_region as RawFd) });
+
+        let virtio_features = self.get_features().unwrap();
+        let total_mmap_size;
+        let queue_region_size;
+        let descr_state_size;
+        if virtio_features & (1 << VIRTIO_F_RING_PACKED) == 0 {
+            // Use descriptor and queue states for split virtqueues
+            queue_region_size = mem::size_of::<QueueRegionSplit>();
+            descr_state_size = mem::size_of::<DescStateSplit>();
+        } else {
+            // Use descriptor and queue states for packed virtqueues
+            queue_region_size = mem::size_of::<QueueRegionPacked>();
+            descr_state_size = mem::size_of::<DescStatePacked>();
+        }
+
+        total_mmap_size = (queue_region_size + descr_state_size * inflight.queue_size as usize)
+            * inflight.num_queues as usize;
+
+        unsafe { libc::ftruncate(inflight_queue_region as RawFd, total_mmap_size as i64) };
+
+        Ok((
+            VhostUserInflight {
+                mmap_size: total_mmap_size as u64,
+                mmap_offset: 0,
+                num_queues: inflight.num_queues,
+                queue_size: inflight.queue_size,
+            },
+            unsafe { File::from_raw_fd(inflight_queue_region as RawFd) },
+        ))
     }
 
-    fn set_inflight_fd(&mut self, _inflight: &vhost::vhost_user::message::VhostUserInflight, _file: File) -> VhostUserResult<()> {
-        Err(VhostUserError::InvalidOperation)
+    fn set_inflight_fd(
+        &mut self,
+        _inflight: &VhostUserInflight,
+        file: File,
+    ) -> VhostUserResult<()> {
+        self.inflight_file = Some(file);
```

Comment:
Same here, please use https://github.com/qemu/qemu/blob/master/subprojects/libvhost-user/libvhost-user.c#L1706 as a reference, and make sure to update the file after you called into `mmap()`.
The previous `inflight_file` will be closed automatically by Rust since you're replacing it with the new file you just received, so we're good there. But you must also `unmap` what has been previously mapped through `get_inflight_fd()`. That's why you need to store something like `inflight_mapping_addr: Option<GuestAddress>` to be able to know if a previous mapping happened.